### DISCUSSION TITLE
how to add a tagfilter?

### DISCUSSION THREAD

**Discussion by Letshadow** [levensailor/ciscoaxl#33] [Q&A] [answered]
Hello again, I am carrying out a project to verify configurations in CUCM.
When I search for a specific user I can access the information very easily, I summarize:

EU=>End User
DN=>Directory number
DV=>Device

![Get Specific](https://user-images.githubusercontent.com/44853747/173251449-3f3dfa25-ebb6-4492-9330-4e82fb2574a2.JPG)

but when I go to do it globally there are many tagfilters that are not enabled,
L=>List
In Red: it is essential for verification
In yellow: not essential
In blue: current method

![Get All](https://user-images.githubusercontent.com/44853747/173251463-fda32d7b-9cbe-4b70-92b1-25ee49adcfd4.JPG)

I understand that it is due to performance problems, the fact is that I know that it can be a long process but I have a powerful machine and I am willing to wait to fill the lists to do a correct debugging on my platform, it would be a monthly process or something similar .

Can you give me instructions to modify the library by enabling the missing tags filter?

In the SOAP I notice that the fields of what I expect to be the list of users (LUser) are specified, would I have to edit it?
```
<xsd:complexType name="LUser">
        <xsd:sequence minOccurs="0">
            <xsd:element maxOccurs="1" minOccurs="0" name="firstName" type="xsd:string"/>
            <xsd:element maxOccurs="1" minOccurs="0" name="middleName" type="xsd:string"/>
            <xsd:element maxOccurs="1" minOccurs="0" firstname="lastname" type="xsd:string"/>
.
.
.
            <xsd:element maxOccurs="1" minOccurs="0" name="user identity" type="xsd:string">
                <xsd:annotation>
                    <xsd:documentation>The value must be in the format user@domain</xsd:documentation>
                </xsd:annotation>
            </xsd:element>
        </xsd:sequence>
        <xsd:attribute name="uuid" type="axlapi:XUUID"/>
    </xsd:complextype>

```
in the axl library I don't know where the fields are enabled, but I know that the zeep library uses them to print the error with enabled fields

```
File ~\AppData\Roaming\Python\Python39\site-packages\zeep\xsd\valueobjects.py:242, at _process_signature(xsd_type, args, kwargs)
    239 available_kwargs.remove("_raw_elements")
    241 if available_kwargs:
--> 242 magnification type error (
    243 ("%s() got an unexpected keyword argument %r. " + "Signature: `%s`")
    244% (
    245 xsd_type.qname or "ComplexType",
    246 next(iter(available_kwargs)),
    247 xsd_type.signature(independent=False),
    248)
    249)
    251 returned result
```

**Comment by rlad78** [comment, marked as answer]
Unfortunately, this is a limitation of AXL, not the library. `get_phones`, `get_users`, and `get_directory_numbers` all use 'list' calls instead of 'get' like their singular counterparts. The [AXL schema](https://developer.cisco.com/docs/axl-schema-reference/) specifies that nearly all 'list' calls must provide a valid 'returnedTags' attribute (this is what `tagfilter` directly assigns to). With 'get' calls, this is optional and AXL will return all tags if 'returnedTags' is not supplied.

There is currently not a way to ask AXL to return all tags on a 'list' call without specifying every single tag manually. Again, the work done in #30 will make this a non-issue, but that is not ready to be implemented yet.

For now, the workaround I would suggest is pulling lists of all users, DN's, and phones but using an empty `tagfilter`. This will only provide a 'uuid' for each item, but you can then use that to search for your user/DN/phone directly. A basic example of this would be:

```py
phones = []
users = []
dns = []

phone_uuids = [phone.uuid for phone in ucm.get_phones(tagfilter={})]
user_uuids = [user.uuid for user in ucm.get_users(tagfilter={})]
dn_uuids = [dn.uuid for dn in ucm.get_directory_numbers(tagfilter={})]

for p_uuid in phone_uuids:
    phones.append(ucm.get_phone(uuid=p_uuid))

for u_uuid in user_uuids:
    users.append(ucm.client.getUser(uuid=u_uuid))  # need to use client here, get_user doesn't have a uuid/args parameter

for d_uuid in dn_uuids:
    dns.append(ucm.get_directory_number(uuid=d_uuid))
```

Depending on how many users/phones/DNs you have however, the multiple 'get' requests can take a LONG time to finish. If you have experience with threading, I would suggest using a [ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor) with a max worker rate of around 50 (depending on how many resources are allocated to your UCM publisher, you could go higher).

**Reply by Letshadow** [reply]
Sounds like a brilliant solution to me, although it also says it will take more time to use the uuid, it may be what I need, I've never used ThreadPoolExecutor but will learn, again thanks for your prompt response,I will wait anxiously to the patch per #30.

I haven't used github much apart from saving things so I'm not very good at comments and stuff, but if you have any boilerplate work or something you think I can help you with, I'd be happy to help. :)
